pysph/cpy/jit.py

```diff
@@ -0,0 +1,523 @@
+                      "Return value defaulting to 'double'", node)
+            self.arg_types['return_'] = 'double'
+
+
+def gather_external_funcs(f_helper, external_f):
+    node_name = f_helper.func.__name__
+    if node_name not in external_f:
+        external_f[node_name] = []
+    for name, child in f_helper.children.items():
+        external_f[node_name].append(child.func)
+        gather_external_funcs(child, external_f)
+
+
+class TranspilerJIT(Transpiler):
+    def __init__(self, backend='cython', incl_cluda=True):
+        """Constructor.
+
+        Parameters
+        ----------
+
+        backend: str: Backend to use.
+            Can be one of 'cython', 'opencl', 'cuda' or 'python'
+        """
+        Transpiler.__init__(self, backend=backend, incl_cluda=incl_cluda)
+
+    def _handle_external(self, func):
+        syms, implicit, calls, externs = get_external_symbols_and_calls(
+            func, self.backend
+        )
+
+        if implicit:
+            msg = ('Warning: the following symbols are implicitly defined.\n'
+                   '  %s\n'
+                   'You may want to explicitly declare/define them.')
+            print(msg)
+
+        self._handle_externs(externs)
+        self._handle_symbols(syms)
+
+        for f in self.external_funcs[func.__name__]:
+            self.add(f)
+
+
+class ElementwiseJIT(parallel.ElementwiseBase):
+    def __init__(self, func, backend='cython'):
+        backend = array.get_backend(backend)
+        self.tp = TranspilerJIT(backend=backend)
+        self.backend = backend
+        self.name = func.__name__
+        self.func = func
+        self._config = get_config()
+        self.cython_gen = CythonGenerator()
+        self.queue = None
+
+    def get_type_info_from_args(self, *args):
+        type_info = {}
+        arg_names = getargspec(self.func)
+        if 'i' in arg_names:
+            arg_names.remove('i')
+            type_info['i'] = 'int'
+        for arg, name in zip(args, arg_names):
+            arg_type = get_ctype_from_arg(arg)
+            if not arg_type:
+                arg_type = 'double'
+            type_info[name] = arg_type
+        return type_info
+
+    @memoize
+    def _generate_kernel(self, *args):
+        if self.func is not None:
+            arg_types = self.get_type_info_from_args(*args)
+            helper = AnnotationHelper(self.func, arg_types)
+            helper.annotate()
+            self.func = helper.func
+            external_funcs = {}
+            gather_external_funcs(helper, external_funcs)
+            self.tp.external_funcs = external_funcs
+        return self._generate()
+
+    def _massage_arg(self, x):
+        if isinstance(x, array.Array):
+            return x.dev
+        elif isinstance(x, np.ndarray):
+            return x
+        else:
+            return np.asarray(x)
+
+    def __call__(self, *args, **kw):
+        c_func = self._generate_kernel(*args)
+        c_args = [self._massage_arg(x) for x in args]
```

Comment:
This is something for later and not this PR but if we are going to call the same jitted function repeatedly, we should find a way to not have to keep massaging the args each time.